File: engine/src/main/java/franca/java/graphics/test_components/TestView0.java
package franca.java.graphics.test_components;

import java.util.Random;

import franca.java.graphics.animations.Ease;
import franca.java.graphics.animations.Tween;
import franca.java.graphics.device.Device;
import franca.java.graphics.views.Page;
import franca.java.graphics.views.View;
import franca.java.graphics.painter.Painter;

public class TestView0 extends View {

  MyTween myTween = null;

  @Override
  public void paint(Device device, Painter painter, Page page) {

    float x = new Random().nextInt(50) + 50f;
    float y = new Random().nextInt(50) + 50f;

    if (myTween != null) {
      painter.paintText(String.valueOf(myTween.getEase().currentValue), x, y, "", 255);
    } else {
      painter.paintText("PAINT", x, y, "", 255);
    }
  }

  @Override
  public void handlePointerDown(Device device, Page page, float painterX, float painterY, float pointedX, float pointedY, int buttonNumber) {
    if (buttonNumber == 1) {
      removeViewAnimation(device);
      myTween = new MyTween(page, this);
      device.registerTween(myTween);
    } else {
      device.requestRepainting();
    }
  }

  public void removeViewAnimation(Device device) {
    if (myTween != null) {
      device.removeTween(myTween);
      myTween = null;
    }
  }

  static class MyTween extends Tween {

    TestView0 testView0;

    MyTween(Page page, TestView0 testView0) {
      super(page, testView0, 500, Ease.EASE_LINEAR);

      this.testView0 = testView0;
    }

    @Override
    public boolean needsRepainting(Device device) {
      boolean result = super.needsRepainting(device);
      if (!result) {
        testView0.removeViewAnimation(device);
        return false;
      }

      return true;
    }
  }
}

File: platforms/typescript/src/test-view-0.ts
import {Tween} from "@java-franca/animations/tween";
import {Page} from "@java-franca/graphics/page";
import {Ease} from "@java-franca/animations/ease";
import {Router} from "@java-franca/expected/router";
import {View} from "@java-franca/graphics/view";
import {Painter} from "@java-franca/expected/painter";

class MyTween extends Tween {
  private testView0: TestView0;

  constructor(page: Page, testView0: TestView0) {
    super(page, null, 500, Ease.EASE_LINEAR);
    this.testView0 = testView0;
  }

  needsRepainting(router: Router): boolean {
    const result = super.needsRepainting(router);
    if (!result) {
      this.testView0.removeViewAnimation(router);
      return false;
    }
    return true;
  }
}

export class TestView0 extends View {
  private myTween: MyTween | null = null;

  paint(router: Router, painter: Painter, page: Page): void {
    const x = Math.floor(Math.random() * 50) + 50;
    const y = Math.floor(Math.random() * 50) + 50;

    if (this.myTween !== null) {
      painter.paintText(
        String(this.myTween.getEase()?.currentValue),
        x,
        y,
        '',
        255
      );
    } else {
      painter.paintText('PAINT', x, y, '', 255);
    }
  }

  handlePointerDown(
    router: Router,
    page: Page,
    painterX: number,
    painterY: number,
    pointedX: number,
    pointedY: number,
    buttonNumber: number
  ): void {
    if (buttonNumber === 1) {
      this.removeViewAnimation(router);
      this.myTween = new MyTween(page, this);
      router.registerTween(this.myTween);
    } else {
      router.requestRepainting();
    }
  }

  removeViewAnimation(router: Router): void {
    if (this.myTween !== null) {
      router.removeTween(this.myTween);
      this.myTween = null;
    }
  }
}
